src/openms/source/KERNEL/MassTrace.cpp

```diff
@@ -283,12 +283,12 @@ namespace OpenMS
     return fwhm_;
   }
 
-    double MassTrace::calculateXCoordinateAtHalfMax_(double xA, double xB, double yA, double yB, double fwhm_int) const
+    double MassTrace::linearInterpolationAtY_(double xA, double xB, double yA, double yB, double y_eval) const
```

Comment:
can you add a `PRECONDITION` to check that `y_ecal` is between `yA` and `yB` to make sure we are interpolating not extrapolating?

Reply:
This preconditions is thrown if no trace point exists beforehand. 
If we do not want to allow this we should decide how to respond: 
Extrapolate or return 0. 

The new algorithm now correctly returns 0 for many sidecases where no value can be calculated. The old algorithm returned 0.'verysmall' values.
If we return 0 we can filter those values out later on.

What do you think should be done in this case?

Reply:
well its quite simple, you cannot interpolate if there is no data point. Again, it is quite hard to think of a case where this would happen but it may happen if the right and left border are exactly equal because a peak is basically a single individual outlier peak. You should probably check for that case because I cannot see any other case where you would want to extrapolate

Reply:
I perfectly agree.
Basically what throws me off is the following:
https://imgur.com/a/BDUYnrq
y_eval is smaller than yA and yB, so no matter what you do with the algorithm, this does not make any sense.
Is the input data already wrong or is there something that I miss with my lacking MS data knowledge?

Reply:
well can you also check whether this code works as expected?

```
     Size left_border(max_idx), right_border(max_idx);
 
     while (left_border > 0 && tmp_ints[left_border] >= half_max_int)
     {
       --left_border;
     }
 
     while (right_border + 1 < tmp_ints.size() && tmp_ints[right_border] >= half_max_int)
     {
       ++right_border;
     }
```

maybe the `max_idx` is already at 1.3e6 and there is simply not data point picked further to the right? Again, in this case we should not extrapolate but (i) fix the mass trace itself or (ii) simply take the rightmost data point. I think for both cases you should simply check that your leftmost / rightmost datapoint is below the half intensity.

Reply:
and return 0 otherwise to indicate that we don't have a proper FWHM?

Reply:
> and return 0 otherwise to indicate that we don't have a proper FWHM?

hmm I would not do that but simply return the leftmost / rightmost point that we have. I dont think zero is a good solution in that case

Reply:
this will often lead to fwhm that are half the size as expected and in my view not much better than extrapolating (or in some cases even worse). But I have no strong feelings about that and we just need to decide if we want to throw them out

Reply:
> this will often lead to fwhm that are half the size as expected and in my view not much better than extrapolating (or in some cases even worse). But I have no strong feelings about that and we just need to decide if we want to throw them out

I disagree, if you only capture half of the elution trace than your FWHM estimate is the least of your worries - also your quantification will be off etc. So unless we also set those to zero I dont think we should set FWHM to zero

Reply:
Yes, I think we are on the same line and that's what I actually wanted to say with "need to decide if we want to throw them out". If we don't have a proper mass trace then we might consider setting FWHM, quality and quantification to 0

Reply:
So should we, in these cases, go the more conservative way of setting everything to 0 to clearly indicate that something went wrong during mass trace extraction / quantification? Or choose something that could be kind of close to the FWHM (but also totally off in some cases).

Reply:
> So should we, in these cases, go the more conservative way of setting everything to 0 to clearly indicate that something went wrong during mass trace extraction / quantification? Or choose something that could be kind of close to the FWHM (but also totally off in some cases).

In principle I agree but I feel that this is the wrong place to discuss because (i) we had this issue of incomplete masstraces being reported before and this is not the PR to address this issue but to actually improve OpenMS at another place and (ii) determining whether to throw out a masstrace should depend on other factors than the FWHM and the user should have control over which traces to throw out. Finally (iii) the best option could be to retrace our steps and improve the masstrace in case we find an issue like this